### PR TITLE
refactor: do not recompute every time num_docs

### PR DESCRIPTION
**Pull request by JoanFM**

### CODE REVIEW

File: docarray/index/backends/hnswlib.py [hunk 1]
```diff
@@ -628,8 +627,11 @@ def accept_hashed_ids(id):
 
         # If hashed_ids is provided, k is the minimum of limit and the length of hashed_ids; else it is limit
         k = min(limit, len(hashed_ids)) if hashed_ids else limit
-
-        labels, distances = index.knn_query(queries, k=k, **extra_kwargs)
+        try:
+            labels, distances = index.knn_query(queries, k=k, **extra_kwargs)
+        except RuntimeError:
+            k = min(k, self.num_docs())
+            labels, distances = index.knn_query(queries, k=k, **extra_kwargs)
```
Comment (jupyterjazz) [inline docarray/index/backends/hnswlib.py:634]: we should just make sure that value of `k` is not exceeding number of vectors, no need for try catch

**Review by jupyterjazz — CHANGES_REQUESTED** [review]
value of `k` should not be more than number of vectors, otherwise lgtm

File: docarray/index/backends/hnswlib.py (hunk 1 above)
Comment (jupyterjazz) [reply docarray/index/backends/hnswlib.py:634]: I already modified the code so you can revert this change

Comment (JoanFM) [reply docarray/index/backends/hnswlib.py:634]: but this would involve having a call to num_docs very often which is what I am trying to avoid

Comment (jupyterjazz) [reply docarray/index/backends/hnswlib.py:634]: why does it matter so much it's cached

Comment (JoanFM) [reply docarray/index/backends/hnswlib.py:634]: I have seen it take 13 seconds for when the index is of the order of 1M documents. Just to solve a couple of very edge cases I do not think is worth

**Review by jupyterjazz — APPROVED** [review]
I wasn't aware of this issue. Let's add 1-2 sentences in the code to explain why we're doing this

**Pull request merged**